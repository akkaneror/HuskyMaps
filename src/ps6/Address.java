package ps6;

/** An Address is an immutable record type defining a location.
 *   @specfield num     : integer // number of the building along the street
 *   @specfield name    : String  // name of the street
 *   @specfield zipcode : String  // US ZIP code
 */
public class Address {

    // AF(x):
    //    num = x.streetNum
    //    name = x.streetName
    //    zipcode = x.zipcode
    // RI:
    //    num >= 0
    //    streetName != null &&
    //    zipcode != null

    private final int streetNum;
    private final String streetName;
    private final String zipcode;

    /**
     * Address Constructor
     *
     * @requires num > 0 && name != null && zipcode != null
     * @effects returns an Address with the given field values
     * */
    public Address (int num, String name, String zipcode) {
        if (num < 0) {
            throw new IllegalArgumentException("Negative street number");
        } else if (name == null) {
            throw new NullPointerException("Street name is null");
        } else if (zipcode == null) {
            throw new NullPointerException("Zip code is null");
        }
        this.streetNum = num;
        this.streetName = name.intern();
        this.zipcode = zipcode.intern();
    }

    /**
     * Get the street number
     * @returns this.num
     */
    public int getNum () {
        return streetNum;
    }

    /**
     * Get the name of the street
     * @returns this.name
     */
    public String getName () {
        return streetName;
    }

    /**
     * Get the zipcode for the street
     * @returns this.zipcode
     */
    public String getZipcode () {
        return zipcode;
    }

    public boolean equals (Object other) {
        return (other instanceof Address) && equals((Address) other);
    }

    public boolean equals (Address other) {
        //can use referential equality because strings are interned
        return (other != null) &&
            zipcode == other.zipcode &&
            streetName == other.streetName &&
            (streetNum == other.streetNum);
    }

    public int hashCode () {
        return streetName.hashCode() + zipcode.hashCode() * 7 + streetNum * 17;
    }

    /**
     * @return a String representation of this address in the format:
     * "num name zipcode"
     **/
    public String toString () {
        return new String (streetNum + " " + streetName + " " + zipcode);
    }
}
