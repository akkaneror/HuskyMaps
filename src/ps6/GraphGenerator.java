package ps6;

import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import ps2.GeoPoint;
import ps4.StreetSegment;
import ps3.graph.Graph;

/**
 * A GraphGenerator is a factory class to produce a Graph of StreetSegments of a geographic area.
 * Geographic information is obtained from a valid Tiger Database Directory, a directory of geographical 
 * street and address information.
 * 
 */
public class GraphGenerator {

	private static boolean DEBUG = false;
	
	/**
	 * Static factory method to produce a graph representing the geographic area 
	 * data accessible by the specified database reader.
	 * @requires segments != null
     * @param Collection segments of StreetSegments containing all segments contained in this graph 
	 * @return a Graph of StreetSegments representing the geographic area.
	 */
	public static Graph<StreetSegment> produceGraph(Collection<StreetSegment> segments){
    	
		// our graph
    	Graph<StreetSegment> geoGraph = new Graph<StreetSegment>();
    	
    	// storage maps
        Map<GeoPoint, Set<StreetSegment> > startPoints = new HashMap<GeoPoint, Set<StreetSegment> >();
        Map<GeoPoint, Set<StreetSegment> > endPoints = new HashMap<GeoPoint, Set<StreetSegment> >();

        long startNodes = System.currentTimeMillis();
        // <<<< add segments to graph and storage maps >>>>
        // for all segments
        for(StreetSegment gs : segments){
        	
    		// add segment to graph
    		geoGraph.addNode(gs);
    		// add segment to storage maps
        	createMapping(startPoints, gs, gs.getP1());
        	createMapping(endPoints, gs, gs.getP2());
        	
        	
        	// add reverse of segment to graph
        	gs = gs.reverse();
        	geoGraph.addNode(gs);
        	// add to storage maps
        	createMapping(startPoints, gs, gs.getP1());
        	createMapping(endPoints, gs, gs.getP2());
        }
        long endNodes = System.currentTimeMillis();
        if(DEBUG) System.out.println("Time for nodes: " + (endNodes-startNodes)/1000.0);
        
        startNodes = System.currentTimeMillis();
        // <<<< connect edges >>>>
        // iterate over end points to add edges to start points
		for(GeoPoint gp : endPoints.keySet()){
		   
			// for all possible children segments of this end point
			Set<StreetSegment> children = startPoints.get(gp);
				
			// for each parent segment with this end point
			for(StreetSegment sP : endPoints.get(gp)){
				
				if(children != null){     // if there exists children
					// get all child segments that start at this parent segment's end point
					for(StreetSegment sC : children){
						// add the edge
						geoGraph.addEdge(sP, sC);
					}
				}
				
			}
        }
		endNodes = System.currentTimeMillis();
        if(DEBUG) System.out.println("Time for edges: " + (endNodes-startNodes)/1000.0);
		return geoGraph;
	}
	

    
    
    /** helper to populate storage maps of points indexed to sets of segments */
    private static void createMapping(Map<GeoPoint, Set<StreetSegment> > m,
    			StreetSegment gs, GeoPoint p){
    	Set<StreetSegment> t = !m.containsKey(p) ? new HashSet<StreetSegment>() : m.get(p);
    	t.add(gs);
    	m.put(p, t);
    }
	
}
