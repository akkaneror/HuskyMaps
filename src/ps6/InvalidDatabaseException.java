package ps6;

/**
 * Checked exception to indicate that there was an error
 * in loading the Tiger database.
 **/
public class InvalidDatabaseException extends Exception {
    public static final long serialVersionUID = 4534;
    public InvalidDatabaseException() {
        super();
    }
    public InvalidDatabaseException (String dbname) {
        super (dbname);
    }
    public InvalidDatabaseException (String dbname, Throwable cause) {
        super(dbname, cause);
    }
}
