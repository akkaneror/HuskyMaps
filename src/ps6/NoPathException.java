package ps6;

/**
 * Checked exception to indicate that a path was not found.
 **/
public class NoPathException extends Exception {
    public static final long serialVersionUID = 4534;
    public NoPathException () {
        super ();
    }
    public NoPathException(String s) {
        super(s);
    }
}
