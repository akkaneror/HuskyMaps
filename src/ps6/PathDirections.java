package ps6;

import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import ps2.Route;

/**
 *   A PathDirections is a description of how to travel from one Address to
 *   another.
 *   @specfield start      : Address // starting point
 *   @specfield end        : Address // destination
 *   @specfield length     : double  // traveling distance, in miles (not rounded)
 *   @specfield route      : Route  // this sequence of segments the directions
 *   								// travel over
 *   @specfield directions : sequence[String] //
 *   directions for traveling from start to end.  Each
 *   String in the sequence must be one line of directions, in the format
 *   specified in <a href="../../psets/ps6/ps6.html#dirs-format">PS6
 *   text based interface</a>, <b>excluding</b> the first line which
 *   states where you start <b>and</b> the final line which gives the length/time
 *   of the trip.  The strings <b>must not</b> include newlines at their
 *   ends.  The order of the Strings in the sequence must correspond to the order
 *   of the lines in the directions from start to end.
 */

public class PathDirections implements Directions{

	/** The start address of these directions */
	private final Address start;
	/** The end address of these directions */
	private final Address end;
	/** The route geographically representing these directions */
	private final Route route;
	/** Sequential list of navigational instructions */
	private final List<String> directions;
	
	/**
	 * Creates a new PathDirections containing the specified sequence of directions 
	 * @requires directions != null && route != null && start != null && end != null
	 * @effects Create a new PathDirections object with the specified list of 
	 * strings as its sequence of directions.
	 */
	public PathDirections(List<String> directions, Route route, Address start, Address end){
		if(directions == null || route == null || start == null || end == null){
			throw new IllegalArgumentException("Specified arguments must not be null");
		}
		
		this.start = start;
		this.end = end;
		this.route = route;
		this.directions = Collections.unmodifiableList(directions);
	}
	
	
    /** 
     * Get the length of these directions, in miles
     * @return this.length. 
     */
    public double getLength (){
    	return route.getLength();
    }

    /** 
     * Get an iterator over this sequence of directions
     * @return an Iterator of Strings over this.directions 
     */
    public Iterator<String> iterator(){
    	return directions.iterator();
    }

    /**
     * Get the start address 
     * @return this.start 
     */
    public Address getStart (){
    	return start;
    }

    /**
     * Get the end address 
     * @return this.end 
     */
    public Address getEnd (){
    	return end;
    }
    
    /**
     * Get the route geographically representing these directions
     * @return this.route 
     */
    public Route getRoute(){
    	return route;
    }
    
    /**
     * @return these directions in a single new-line-separated string 
     */
    public String toString(){
    	String result = "Start at "+start+"\n";
		
		Iterator<String> iter = iterator();
		while(iter.hasNext()){
			result += iter.next()+"\n";
		}
		
		return result + "Trip length: "+route.getLength();
    }
	
}
