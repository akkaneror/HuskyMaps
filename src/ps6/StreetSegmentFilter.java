package ps6;

import ps4.*;

/**
 * Interface for Street Segment filtering
 */
public interface StreetSegmentFilter {

  /**
   *  Determine if the segment is accepted by the filter
   *
   *  @param seg a StreetSegment candidate
   *  @return true iff the filter accepts seg
   */
  public boolean apply(StreetSegment seg);
}
