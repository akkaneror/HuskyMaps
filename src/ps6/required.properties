required.files = README,DirectionsFinder.java,TextUI.java,test/SpecificationTests.java,test/ImplementationTests.java,answers/gui.txt,answers/reflection.txt,answers/collaboration.txt