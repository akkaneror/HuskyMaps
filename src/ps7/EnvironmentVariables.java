package ps7;

/**
 * Class of static variables and switches that control the behavior of extra
 * features of HuskyMaps.  Accessed by TextUI and GraphicUI with command line arguments.
 */

public class EnvironmentVariables {

	/** Switch to turn on fractional segments */
	public static boolean FRACTIONAL_SEGMENTS_ON =  false;
	
	/** Switch to turn on Pathfinding algorithm for performance.  Doubles the cost
	 * of A*'s hypothetical distance-to-goal pathfinding optimization. */
	public static boolean OPTIMIZE_FOR_PERFORMANCE_ON = false;

	/** Switch to turn on intersection traversal cost */
	public static boolean INTERSECTION_TRAVERSAL_ON = false;
	/** Standard measure of the cost of traversing an intersection, in miles. 
	 * Based on an average wait time of a FIFTEEN SECONDS at each intersection.
	 * If street classification turned on, based on travel time for the street class.
	 * Otherwise, FIFTEEN SECONDS wait time calculated for an average speed of 2 minutes-per-mile. 
	 * ( 30 MPH ) therefore, .25 miles-per-minute intersection cost */
	public static final double INTERSECTION_TRAVERSAL_COST = .25;
	/** Standard number of intersecting streets required to be considered an intersection. */
	public static final int STREETS_PER_INTERSECTION = 3;
	
	/** Switch to turn on street classification cost */
	public static boolean STREET_CLASSIFICATION_ON = false;
	/** Modifier of based on Street Classification, representing standard 
	 * minutes-per-mile averages for primary highway road type. */
	public static final double PRIM_HWY_COST = 60.0/60.0;
	/** Modifier of based on Street Classification, representing standard 
	 * minutes-per-mile averages for secondary highway road type. */
	public static final double SEC_HWY_COST = 60.0/35.0;
	/** Modifier of based on Street Classification, representing standard 
	 * minutes-per-mile averages for local road type. */
	public static final double LOCAL_ROAD_COST = 60.0/25.0;
	/** Modifier of based on Street Classification, representing standard 
	 * minutes-per-mile averages for an unknown road type. */
	public static final double UNKNOWN_ROAD_COST = 60.0/15.0;
	
    /** Factor for modified A* of cost = path length + (distance to nearest goal) * FACTOR 
     *  Increasing factor improves PathFinding operating time with a corresponding loss in accuracy.
     *  Meant to be used in conjunction with the MAXIMUM_BUFFER_DISTANCE.  This PathFinding
     *  optimization will be active if intersection traversal is turned on. */
	public static final double DISTANCE_TO_GOAL_COST_FACTOR = 3.00;
	/** Maximum buffer distance in miles surrounding the leading edge of the most probable path.
	 * Within this buffer distance, path distance-to-goal cost ( of the A* path finding algorithm )
	 * is determined normally.  Outside this distance, path distance-to-goal cost should be
	 * exponentially increased.  This PathFinding optimization will be active if  
	 * intersection traversal is turned on. */
	public static double MAXIMUM_BUFFER_DISTANCE = 0.5;
		
	
	/** Processes the specified command line flag and sets the corresponding switch */
	public static void setFlag(String flag){

		if(flag.equals("-F")){
			FRACTIONAL_SEGMENTS_ON =  true;
		}else if(flag.equals("-f")){
			FRACTIONAL_SEGMENTS_ON =  false;
		}else if(flag.equals("-I")){
			INTERSECTION_TRAVERSAL_ON = true;
		}else if(flag.equals("-i")){
			INTERSECTION_TRAVERSAL_ON = false;
		}else if(flag.equals("-C")){
			STREET_CLASSIFICATION_ON = true;
		}else if(flag.equals("-c")){
			STREET_CLASSIFICATION_ON = false;
		}
		
	}
	
}
