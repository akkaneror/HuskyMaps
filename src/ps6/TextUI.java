package ps6;

import java.io.*;
import java.text.DecimalFormat;
import java.util.*;

import ps6.Address;
import ps7.EnvironmentVariables;
import ps2.*;
/**
 * This class implements an interactive text-based interface for Husky Maps.
 * Valid command line arguments can displayed with the single flag '-h'. 
 **/

public class TextUI {
	
	/** switch to turn on status reporting for the class during runtime, printed to the console */
	public static boolean VERBOSE = false;

	
    /**
	* @param args, for each element of args, the following flags are valid:</br>
	* -h for help text
	* -F to turn on fractional distance, -f to turn off
	* -I to turn on intersection traversal, -i to turn off
	* -C to turn on street classification, -c to turn off
	* after any flags, the first argument is the database path, while all
	* arguments after that are zipcodes to filter over, if any.
	*/
    public static void main(String args[]) {
        try {
            if (args.length == 0) {
                printUsage();
                return;
            }

            TextUI textui;

            int index = 0;
            String next = args[index];
            while(next.startsWith("-")){
                if(args[index].equals("-h")){
        			System.err.println("Set HuskyMaps environment variables!\n" +
        					"USAGE:\n" +
        					"-F to turn on fractional distance, -f to turn off\n" +
        					"-I to turn on intersection traversal, -i to turn off\n" +
        					"-C to turn on street classification, -c to turn off");
        			return;
        		}else{
    				EnvironmentVariables.setFlag(args[index]);
        		}
                next = args[index];
                index++;
            }
            String databaseName = next;
            List<String> zipcodes = new ArrayList<String>();
            for (int i = 1; i<args.length; i++) {
                zipcodes.add(args[i]);
            }

            textui = new TextUI(databaseName, zipcodes);


            BufferedReader r = new BufferedReader(new InputStreamReader(System.in));
            textui.interactiveLoop(r, System.out);

        } catch (IOException e) {
            System.out.flush();
            System.err.println(e.toString());
            e.printStackTrace(System.err);

        }
    }

    private static void printUsage() {
        System.err.println("Usage:");
        System.err.println("java ps6.TextUI [flags] directory-name  zipcode1 zipcode2 ...  zipcodeN\n" +
        		"Type ps6.TextUI -h for valid flags");
    }

    private final String databaseName;
    private final List<String> zipcodes;

    /**
     * @requires databaseName != null && zipcodes != null && 
     * !zipcodes.contains(invalid zipcode format)
     * A valid zipcode is defined as a String that contains exactly 5 numerals, from 0 to 9. 
     *
     * @effects Creates a new TextUI which loads from databaseName
     * and limits its search to the zipcodes given, unless the list
     * is empty or null, in which case it searches over all zipcodes.
     * @throws IllegalArgumentException if a passed zipcode is invalid
     **/
    public TextUI(String databaseName, List<String> zipcodes) {
        this.databaseName = databaseName;
        this.zipcodes = new ArrayList<String>(zipcodes);
        
        for(String s : zipcodes){
        	if(s.length() != 5){
        		throw new IllegalArgumentException("zipcode must be a 5 digit number");
        	}
        	try{
        		Integer.parseInt(s);
        	}catch(NumberFormatException ex){
        		throw new IllegalArgumentException("zipcode must be a 5 digit number");
        	}
        }
    }

    /**
     * Generates text-based user interaction prompts to ask the user for and starting 
     * point and desired destination.  It prints out the shortest-distance directions 
     * from the start point to the destination as a series of navigation instructions.
     *  An example of the instructions format: 
     * 
     * 		"Turn sharp left onto MacGregor Street and [go xxx distance]."
     * 
     * Where the format of '[go xxx distance]' depends on the specified mode of 
     * transportation.
     * 
     * To terminate the interaction loop, specify the "starting number" (at the
     * beginning of the prompt) as '-1'.
     * 
     * @throws InvalidDatabaseException if specified source database is invalid
     **/
    public void interactiveLoop(BufferedReader input, PrintStream output)
        throws IOException
    {
    	
    	if(VERBOSE){ System.out.println("--> Constructing the DirectionsFinder ..."); }
    	
    	DirectionsFinder virtualGPS = null;
    	try{
    		virtualGPS = DirectionsFinder.getDirectionsFinder(databaseName, zipcodes);
    		virtualGPS.VERBOSE = VERBOSE;
    		//throws InvalidDatabaseException if specified source database is invalid
    		
    	}catch(Exception ex){
    		output.println("Database error");
//            System.err.println(ex.toString());
//            ex.printStackTrace(System.err);
            
    	}finally{
    		    		
    		// UI loop:
    		do{
	            //   read addresses from input
	    		Address start = getAddress(input, output, true);
	    		if(start == null){  // if the starting address was specified as the exit command
	    			break;
	    		}
	    		Address end = getAddress(input, output, false);
	    		if(end == null){
	    			throw new RuntimeException("end address should never be set to null");
	    		}

	    		// get travel format
	        	output.print("walking or driving [w/d]? ");
	        	String travelMode = input.readLine().trim();
		    	RouteFormatter formatter = setFormat(travelMode);
		    	
		    	//////////// prompting complete ////////////
		    	
		    	if(VERBOSE){ System.out.println("--> Checking the addresses are valid ..."); }
		    	
		    	// address and directions format checks
		    	int startValid = virtualGPS.checkAddress(start);
		    	int endValid = virtualGPS.checkAddress(end);
		    	
		    	if(startValid != DirectionsFinder.ADDRESS_VALID){       // if start address is bad
		    		output.println(addressError(start, startValid));
		    		break;
		    	}else if(endValid != DirectionsFinder.ADDRESS_VALID){       // if end address is bad
		    		output.println(addressError(end, endValid));
		    		break;
		    	}else if(formatter == null){     // if directions format is invalid
		    		output.println("Invalid direction type: "+travelMode);
		    		break;
		    	}else{

		    		///////////// valid input provided ////////////
		    		
		        	if(VERBOSE){ System.out.println("--> Calculating directions ..."); }
		    		
		    		// attempt to find a path
		        	Directions navigate = null;
		        	try{
		        		navigate = virtualGPS.getDirections(start, end, formatter);
		        		
		        	}catch(InvalidAddressException ex){     // handle errors of bad input
		        		// this error should be caught earlier
		        		output.println("DirectionsFinder address indexing inconsistant");
		            //    System.err.println(ex.toString());
		            //    ex.printStackTrace(System.err);
		        		
		        	}catch(NoPathException ex){     // no path found
		        		output.println("No path from "+start+" to "+end);
		        		
		        	}finally{
		        	
		        		/////////////// valid directions found ////////////
		    		
		            	if(VERBOSE){ System.out.println("--> Displaying the results ..."); }
		        		
		        		displayDirections(navigate, output, formatter);
		        	}
		    	}
		    	
    		}while(true);
	    	
    	}
    }
    
    
    /** handles and outputs Directions as calculated by DirectionsFinder */
    private void displayDirections(Directions navigate, PrintStream output, RouteFormatter formatter)
    {
    	if(navigate == null){
    		throw new IllegalArgumentException("directions object should not be null");
    	}
    	
		//   write to output
		output.println("Start at "+navigate.getStart());
		
		Iterator<String> iter = navigate.iterator();
		while(iter.hasNext()){
			output.println(iter.next());
		}
		
		//distinguish between miles and minutes
		if(formatter instanceof WalkingRouteFormatter){
			//output.println("Trip time: " + (Math.round(navigate.getLength()) * 20) + " minutes");
			
			//output.print("Trip time: "+Math.round(navigate.getLength()*20)+" minutes\n");
			output.printf("Trip time: %.0f minutes\n", navigate.getLength() * 20);
		}else if(formatter instanceof DrivingRouteFormatter){
			//DecimalFormat printer = new DecimalFormat("#.#"); // <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
			
			//output.print("Trip length: "+(Math.round(navigate.getLength()*10))/10.0+" miles\n");
			output.printf("Trip length: %.1f miles\n", navigate.getLength());
		}else{
			throw new IllegalArgumentException("Directions implementor class not recognized");
		}
    }
    	
    
    
    /** helper to prompt user for information about an address and return an Address object 
     * @return an Address, or null if the user specifies "-1" as an exit command */
    private Address getAddress(BufferedReader input, PrintStream output, boolean starting)
    		throws IOException
    {
    	String where = (starting ? "starting" : "destination");
    	
    	output.print( where +" number? ");
    	String address = input.readLine().trim();
    	if(starting && address.equals("-1")){  // if the starting number is the exit command
    		return null;
    	}
    	
    	output.print( where +" street? ");
    	String street = input.readLine().trim();
    	output.print( where +" zipcode? ");
    	String zip = input.readLine().trim();

    	return new Address(Integer.parseInt(address), street, zip);
    }
    
    /** helper to prompt user for travel method and return a RouteFormatter for that method 
     * @return a RouteFormatter, or null if an invalid format is specified*/
    private RouteFormatter setFormat(String mode)
    {
    	if(mode.equals("w")){
    		return new WalkingRouteFormatter();
    	}else if(mode.equals("d")){
    		return new DrivingRouteFormatter();
    	}else{    // invalid argument
    		return null;
    	}
    }
    
    /** helper to handle error-reporting 
     * @return a String of the error report */
    private String addressError(Address addr, int cause){

    	if(cause == DirectionsFinder.ADDRESS_VALID){
    		return "Should not have thrown error for "+addr;
    	}else if(cause == DirectionsFinder.ADDRESS_ZIPCODE_INVALID){
			return "No such zipcode: "+addr;
		}else if(cause == DirectionsFinder.ADDRESS_STREET_INVALID){
			return "No such street: "+addr;
		}else if(cause == DirectionsFinder.ADDRESS_NUM_INVALID){
			return "No such number: "+addr;
		}else{  // address is null
			return "Invalid address argument";
		}
    }
    
	
	
}
