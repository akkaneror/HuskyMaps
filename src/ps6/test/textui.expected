starting number? starting street? starting zipcode? destination number? destination street? destination zipcode? walking or driving [w/d]? Start at 788 NW Alexander St 98239
Turn right onto NW Alexander St and go 0.1 miles.
Turn left onto 7th St NW and go 0.1 miles.
Turn right onto Main St N and go 0.2 miles.
302 Main St N 98239 is on your right
Trip length: 0.4 miles
starting number? starting street? starting zipcode? destination number? destination street? destination zipcode? walking or driving [w/d]? Start at 1240 Arnold Rd 98277
Turn left onto Arnold Rd and walk for 2 minutes.
Turn right onto Monroe Landing Rd and walk for 8 minutes.
Turn left onto Scenic Heights Rd and walk for 5 minutes.
519 Scenic Heights Rd 98277 is on your left
Trip time: 15 minutes
starting number? 
